core/shared/src/main/scala/scalaz/zio/stm/STM.scala

```diff
@@ -0,0 +1,584 @@
+   */
+  final def withFilter(f: A => Boolean): STM[E, A] = filter(f)
+
+  /**
+   * Sequentially zips this value with the specified one.
+   */
+  final def zip[E1 >: E, B](that: => STM[E1, B]): STM[E1, (A, B)] =
+    (self zipWith that)((a, b) => a -> b)
+
+  /**
+   * Sequentially zips this value with the specified one, discarding the
+   * second element of the tuple.
+   */
+  final def zipLeft[E1 >: E, B](that: => STM[E1, B]): STM[E1, A] =
+    (self zip that) map (_._1)
+
+  /**
+   * Sequentially zips this value with the specified one, discarding the
+   * first element of the tuple.
+   */
+  final def zipRight[E1 >: E, B](that: => STM[E1, B]): STM[E1, B] =
+    (self zip that) map (_._2)
+
+  /**
+   * Sequentially zips this value with the specified one, combining the values
+   * using the specified combiner function.
+   */
+  final def zipWith[E1 >: E, B, C](that: => STM[E1, B])(f: (A, B) => C): STM[E1, C] =
+    self flatMap (a => that map (b => f(a, b)))
+}
+
+object STM {
+
+  private[stm] object internal {
+    class Versioned[A](val value: A)
+
+    type Journal =
+      MutableMap[Long, STM.internal.Entry]
+
+    type Todo = () => Unit
+
+    /**
+     * Resets the journal so that it does not modify any entries.
+     */
+    final def resetJournal(journal: Journal): Unit =
+      journal.values foreach (_.reset())
+
+    /**
+     * Atomically collects and clears all the todos from any `TRef` that
+     * participated in the transaction. This is not a pure function, despite
+     * the return type (it effectfully clears todos from `TRef` values).
+     */
+    final def collectTodos(trefs: Iterable[TRef[_]]): Iterable[Todo] = {
+      val allTodos  = MutableMap.empty[Long, Todo]
+      val emptyTodo = Map.empty[Long, Todo]
+
+      trefs foreach { tref =>
+        val todo = tref.todo
+
+        var loop = true
+        while (loop) {
+          val oldTodo = todo.get
+
+          loop = !todo.compareAndSet(oldTodo, emptyTodo)
+
+          if (!loop) {
+            allTodos ++= oldTodo
+          }
+        }
+      }
+
+      allTodos.values
+    }
+
+    /**
+     * For the given transaction id, adds the specified todo effect to all
+     * `TRef` values.
+     */
+    final def addTodo(txnId: Long, trefs: Iterable[TRef[_]], todoEffect: Todo): Unit =
+      trefs foreach { tref =>
+        var loop = true
+        while (loop) {
+          val oldTodo = tref.todo.get
+
+          val newTodo = oldTodo updated (txnId, todoEffect)
+
+          loop = !tref.todo.compareAndSet(oldTodo, newTodo)
+        }
+      }
+
+    final val succeedUnit: TRez[Nothing, Unit] =
+      TRez.Succeed(())
+
+    final def makeTRefId(): Long = trefCounter.incrementAndGet()
+
+    final def makeTxnId(): Long = txnCounter.incrementAndGet()
+
+    private[this] val trefCounter: AtomicLong = new AtomicLong()
+
+    private[this] val txnCounter: AtomicLong = new AtomicLong()
+
+    final val globalLock = new java.util.concurrent.Semaphore(1)
```

Comment:
What would it take to chop this up? Can transactions lock on all participating `TRef`s instead?

Reply:
It's certainly possible. First I want to squeeze down the overhead in this implementation and benchmark the solution. The amount of work being done in the lock is extremely minimal (a check followed by commit), so when optimized, it's possible it will achieve higher throughput than a more complicated, fine-grained strategy, at least for some number of cores `n` (obviously as `n` increases, the overhead becomes less important; figuring out if the cutoff is `4` or `40` seems important).